src/mlpack/methods/ann/rnn_impl.hpp

```diff
@@ -164,9 +164,16 @@ void RNN<OutputLayerType, InitializationRuleType, CustomLayers...>::Predict(
       Forward(std::move(arma::mat(predictors.slice(seqNum).colptr(begin),
           predictors.n_rows, effectiveBatchSize, false, true)));
 
-      results.slice(seqNum).submat(0, begin, results.n_rows - 1, begin +
-          effectiveBatchSize - 1) = boost::apply_visitor(outputParameterVisitor,
+      arma::mat out = boost::apply_visitor(outputParameterVisitor,
           network.back());
+
+      if (results.n_rows == 0)
+      {
+        results.set_size(out.n_rows, predictors.n_cols, rho);
```

Comment:
I am not in strong favour of this change. We already check and set `outputSize` inside `EvaluateWithGradient` and `Evaluate` methods and adding it inside `Predict` makes it redundant in the code + it is inefficient to have it within 2 for loops here. 
Rather I think we could set `outputSize` once inside the `ResetParameters()` method by creating a `OutputSizeVisitor` and reverse iterating over the layers until we have a non-zero value. This way we can remove the need to check `outputSize == 0` inside every method.
I think we are doing something similar here https://github.com/mlpack/mlpack/pull/1366 which has an implementation of the `outputSizeVisitor`.
Let me know what you think. @MuLx10 @zoq

Reply:
@saksham189 I agree that checking redundant conditions within two loops is very inefficient. I will rectify that. However, I am not sure that I should remove that from other functions like Evaluate().  @zoq I would like to have your views.

Reply:
Sounds like a good idea to me, @MuLx10 feel free to remove the part from the other functions as well.

Reply:
okay sure